src/query/service/src/interpreters/interpreter_select_v2.rs

```diff
@@ -81,7 +149,31 @@ impl Interpreter for SelectInterpreterV2 {
     /// The QueryPipelineBuilder will use the optimized plan to generate a Pipeline
     #[tracing::instrument(level = "debug", name = "select_interpreter_v2_execute", skip(self), fields(ctx.id = self.ctx.get_id().as_str()))]
     async fn execute2(&self) -> Result<PipelineBuildResult> {
-        let build_res = self.build_pipeline().await?;
-        Ok(build_res)
+        // Only single table query can use result cache now. Multi-tables query will be supported later.
+        if self.ctx.get_settings().get_enable_query_result_cache()?
+            && self.metadata.read().tables().len() == 1
+        {
+            // 0. Need to build pipeline first to get the partitions.
+            let mut build_res = self.build_pipeline().await?;
+
+            // 1. Try to get result from cache.
+            let kv_store = UserApiProvider::instance().get_meta_store_client();
+            let cache_reader = ResultCacheReader::create(
+                self.ctx.clone(),
+                kv_store.clone(),
+                self.ctx
+                    .get_settings()
+                    .get_tolerate_inconsistent_result_cache()?,
+            );
+            if let Some(blocks) = cache_reader.try_read_cached_result().await? {
```

Comment:
If `try_read_cached_result` returns an error, we can't raise it to the user. Instead, we should log the error and continue the query to make it work.